### PR TITLE
expose nginx ingress

### PR DESCRIPTION
**Pull request by yagonobre**
This change is [Reviewable](https://reviewable.io/reviews/luizalabs/teresa/360)

### CODE REVIEW

File: pkg/protobuf/app/app.proto [hunk 1]
```diff
@@ -17,6 +17,7 @@ message CreateRequest {
     string name = 1;
     string team = 2;
     string process_type = 3;
+  	string host = 6;
```
Comment (aguerra) [inline pkg/protobuf/app/app.proto]: Minor indentation problem

File: pkg/server/deploy/deploy_test.go [hunk 2]
```diff
@@ -186,38 +181,33 @@ func TestCreateDeployReturnError(t *testing.T) {
...
+		{app.ProcessTypeWeb, nil, true},
+		{app.ProcessTypeWeb, errors.New("some sad error"), true},
+		{"worker", nil, false},
 	}
 
 	for _, tc := range testCases {
 		fakeK8s := &fakeK8sOperations{
-			hasSrvReturn: tc.hasSrvReturn,
-			hasSrvErr:    tc.hasSrvErr,
+			hasSrvErr: tc.hasSrvErr,
```
Comment (aguerra) [inline pkg/server/deploy/deploy_test.go]: Minor indent problem

File: pkg/server/k8s/client.go [hunk 3]
```diff
@@ -492,6 +494,59 @@ func (k *k8sClient) CreateService(namespace, appName string) error {
...
+		fmt.Fprintln(w, "Exposing service")
+		if err := k.createService(namespace, appName); err != nil {
+			return err
+		}
+	}
+
+	if k.ingress {
+		hasIgs, err := k.hasIngress(namespace, appName)
+		if err != nil {
+			return nil
```
Comment (drgarcia1986) [inline pkg/server/k8s/client.go]:
```diff
- return nil
+ return err
```

File: pkg/server/k8s/client.go [hunk 4]
```diff
@@ -492,6 +494,59 @@ func (k *k8sClient) CreateService(namespace, appName string) error {
...
+		return err
+	}
+	if !hasSrv {
+		fmt.Fprintln(w, "Exposing service")
+		if err := k.createService(namespace, appName); err != nil {
+			return err
+		}
+	}
+
+	if k.ingress {
```
Comment (drgarcia1986) [inline pkg/server/k8s/client.go]:
```diff
- if k.ingress {
+ if !k.ingress{
+     return nil
+ }
```

File: pkg/server/k8s/converters.go [hunk 5]
```diff
@@ -229,3 +230,36 @@ func serviceSpec(namespace, name, srvType string) *k8sv1.Service {
 		},
 	}
 }
+
+func ingressSpec(namespace, name, host string) *k8s_extensions.Ingress {
```
Comment (drgarcia1986) [inline pkg/server/k8s/converters.go]: what about tests for this func?

File: pkg/server/k8s/client.go [hunk 6]
```diff
@@ -492,6 +494,59 @@ func (k *k8sClient) CreateService(namespace, appName string) error {
...
+	kc, err := k.buildClient()
+	if err != nil {
+		return err
+	}
+	igsSpec := ingressSpec(namespace, appName, host)
+	_, err = kc.ExtensionsV1beta1().Ingresses(namespace).Create(igsSpec)
+	return errors.Wrap(err, "create ingress failed")
+}
+
+// ExposeApp create a service and/or a ingress if needed
```
Comment (drgarcia1986) [inline pkg/server/k8s/client.go]: maybe:
```diff
- create
+ creates
```
? 🤔

File: pkg/server/k8s/converters.go (hunk 5 above)
Comment (yagonobre) [reply pkg/server/k8s/converters.go]: I'll do it later

File: pkg/server/k8s/client.go (hunk 4 above)
Comment (yagonobre) [reply pkg/server/k8s/client.go]: I'm not sure about that, but I no like this approach

Comment (drgarcia1986) [reply pkg/server/k8s/client.go]: why not? with this minor change you can decrease one level of indentation

File: pkg/server/k8s/converters.go (hunk 5 above)
Comment (yagonobre) [reply pkg/server/k8s/converters.go]: done

File: pkg/server/k8s/client.go (hunk 4 above)
Comment (yagonobre) [reply pkg/server/k8s/client.go]: looks unnecessary, I think that this way is more cohesive

**Comment by yagonobre** [comment]
@drgarcia1986 done

File: pkg/server/k8s/client.go (hunk 4 above)
Comment (drgarcia1986) [reply pkg/server/k8s/client.go]: I disagree, but I like to listen to @aguerra opinion.

Comment (aguerra) [reply pkg/server/k8s/client.go]: It's standard idiom in Go to treat the default case as not indented.

Comment (yagonobre) [reply pkg/server/k8s/client.go]: updated.

File: CHANGELOG.md [hunk 7]
```diff
@@ -2,9 +2,13 @@
 All notable changes to this project will be documented in this file.
 This project adheres to [Semantic Versioning](http://semver.org/).
 
-## [Next Version] - Release Date
 ### Added
+=======
```
Comment (drgarcia1986) [inline CHANGELOG.md]: 🤔

Comment (yagonobre) [reply CHANGELOG.md]: 🏃

Comment (yagonobre) [reply CHANGELOG.md]: 🤦‍♂️

**Comment by yagonobre** [comment]
@aguerra @drgarcia1986 rebased

**Comment by drgarcia1986** [comment]
Is there a reason why the label `don't-merge` still here?

**Comment by yagonobre** [comment]
@aguerra @drgarcia1986 @arnaldopereira I change from `host` to `virtual host`

File: pkg/protobuf/app/app.proto [hunk 8]
```diff
@@ -17,6 +17,7 @@ message CreateRequest {
     string name = 1;
     string team = 2;
     string process_type = 3;
+    string virtual_host = 6;
```
Comment (drgarcia1986) [inline pkg/protobuf/app/app.proto]: maybe? :
```diff
- 6
+ 4
```
🤔

File: pkg/server/deploy/deploy.go [hunk 9]
```diff
@@ -28,8 +28,7 @@ type Operations interface {
 type K8sOperations interface {
 	PodRun(podSpec *PodSpec) (io.ReadCloser, <-chan int, error)
 	CreateOrUpdateDeploy(deploySpec *DeploySpec) error
-	HasService(namespace, name string) (bool, error)
-	CreateService(namespace, name string) error
+	ExposeApp(namespace, name, vHost string, w io.Writer) error
```
Comment (drgarcia1986) [inline pkg/server/deploy/deploy.go]: I know it's a bit late, but, what about `ExposeDeploy`? `K8sOperations` care about k8s stuff and `app` terminology is something related to Teresa

File: pkg/protobuf/app/app.proto (hunk 8 above)
Comment (yagonobre) [reply pkg/protobuf/app/app.proto]: change numeric tags for any existing fields may break backward compatibility

File: pkg/server/deploy/deploy.go (hunk 9 above)
Comment (yagonobre) [reply pkg/server/deploy/deploy.go]: Make sense

File: pkg/protobuf/app/app.proto (hunk 8 above)
Comment (drgarcia1986) [reply pkg/protobuf/app/app.proto]: but `virtual_host` is a new field, right?

File: pkg/server/deploy/deploy.go (hunk 9 above)
Comment (yagonobre) [reply pkg/server/deploy/deploy.go]: 🔁 , I change `deployOps.exposeService` to `deployOps.exposeApp` and `k8sOps.ExposeApp` to `k8sOps.ExposeDeploy`. c/c @aguerra

Comment (drgarcia1986) [reply pkg/server/deploy/deploy.go]: Great

File: pkg/protobuf/app/app.proto (hunk 8 above)
Comment (yagonobre) [reply pkg/protobuf/app/app.proto]: yes but fields 4 and 5 already exists

Comment (drgarcia1986) [reply pkg/protobuf/app/app.proto]: you're right :/ sorry

**Pull request merged**